libraries/Microsoft.Bot.Connector/Authentication/AdalAuthenticator.cs

```diff
@@ -0,0 +1,188 @@
+        // Cached requests take less than 0.1 millisecond to resolve, so the semaphore doesn't hurt performance under load tests
+        // unless we have more than 10,000 requests per second, but in that case other things would break first.
+        private static Semaphore tokenRefreshSemaphore = new Semaphore(1, 1);
+        private static readonly TimeSpan SemaphoreTimeout = TimeSpan.FromSeconds(10);
+
+        // Depending on the responses we get from the service, we update a shared retry policy with the RetryAfter header
+        // from the HTTP 429 we receive.
+        // When everything seems to be OK, this retry policy will be empty.
+        // The reason for this is that if a request gets throttled, even if we wait to retry that, another thread will try again right away.
+        // With the shared retry policy, if a request gets throttled, we know that other threads have to wait as well.
+        // This variable is guarded by the authContextSemaphore semphore. Don't modify it outside of the semaphore scope.
+        private static volatile RetryParams currentRetryPolicy;
+
+        private readonly ClientCredential clientCredential;
+        private readonly IOAuthConfiguration oAuthConfig;
+
+        private const string msalTemporarilyUnavailable = "temporarily_unavailable";
+
+        public AdalAuthenticator(ClientCredential clientCredential, IOAuthConfiguration oAuthConfig)
+        {
+            this.oAuthConfig = oAuthConfig;
+            this.clientCredential = clientCredential ?? throw new ArgumentNullException(nameof(clientCredential));
+            this.authContext = new AuthenticationContext(oAuthConfig.Authority);
+        }
+
+        public async Task<AuthenticationResult> GetTokenAsync(bool forceRefresh = false)
+        {
+            return await Retry.Run(
+                task: () => AcquireTokenAsync(forceRefresh),
+                retryExceptionHandler: (ex, ct) => HandleAdalException(ex, ct)).ConfigureAwait(false);
+        }
+
+        private async Task<AuthenticationResult> AcquireTokenAsync(bool forceRefresh = false)
+        {
+            bool acquired = false;
+
+            if (forceRefresh)
+            {
+                authContext.TokenCache.Clear();
+            }
+
+            try
+            {
+                // The ADAL client team recommends limiting concurrency of calls. When the Token is in cache there is never 
+                // contention on this semaphore, but when tokens expire there is some. However, after measuring performance
+                // with and without the semaphore (and different configs for the semaphore), not limiting concurrency actually
+                // results in higher response times overall. Without the use of this semaphore calls to AcquireTokenAsync can take up
+                // to 5 seconds under high concurrency scenarios.
+                acquired = tokenRefreshSemaphore.WaitOne(SemaphoreTimeout);
```

Comment:
per this doc, you don't need the acquired code:
https://docs.microsoft.com/en-us/dotnet/api/system.threading.waithandle.waitone?view=netframework-4.7.2

IF the method returns (aka: does not time out), you have the semaphore. Otherwise it's exceptions. 

You can remove all the acquired code in the try/finally. Likewise, the try/finally should move to just after this method completes, so it's: Acquire try{} finally{release};